mdb/mdb_shell.py

```diff
@@ -31,10 +31,54 @@
     from .mdb_attach import Prog_opts
     from .mdb_client import Client
 
+INTERACT_ESCAPE = chr(29)  # ctrl+]
+INTERACT_CANCEL = chr(3)  # ctrl+c
+INTERACT_EOF = chr(4)  # ctrl+d
 GDBPROMPT = r"\(gdb\)"
 plt.style.use("dark_background")
 
 
+def buffered_input_filter(
+    handle_input: Callable[[str], str]
+) -> Callable[[bytes], bytes]:
+    """Wrap functions to generate arguments for `pexpect.interact` filters.
+
+    Wraps a callback function with a buffer so that instead of receiving each
+    character as it is typed, the filter function is given the current command
+    string. The wrapper also augments/overrides the interaction with common
+    substitutions for control characters:
+    - `Ctrl-d` will send `INTERACT_ESCAPE_CHARACTER` to end the interaction
+      instead of `quit` to the gdb shell.
+
+    The `handle_input` argument is only called after each newline or carriage
+    return, and should return the modified input to be sent to the shell.
+
+    Warning:
+        If `handle_input` is to replace the current input, it must also include
+        the backspace characters needed to remove the current string.
```

Comment:
Hi @fjebaker thanks for the lovely PR. I really appreciate the clear comments / documentation :ok_hand: 

This is the only part I don't fully understand. Is this a future warning if we want to change the functionality of `handle_input` e.g., to modify the text "in-place"?

Reply:
Well at the moment the wrapper will just faithfully send anything truthy over to the gdb shell, which means in theory if `handle_input` was given `hello` and returned `world`, the gdb shell would display `helloworld`. This might be desired, but if instead `handle_input` is to replace the `hello` with `world`, it would need to return  `(BACKSPACE * 5) + "world"`.

Maybe this would be better to put as a "Note" instead of a "Warning"?

Reply:
that's great, thank you